### DISCUSSION TITLE
How to get RAW Metadata byte[]

### DISCUSSION THREAD

**Discussion by Unreal-System** [Zeugma440/atldotnet#154] [Q&A] [answered]
This question is similar to it
https://stackoverflow.com/questions/22695723/create-valid-wav-file-header-for-streams-in-memory/63292656#63292656

Is there a method like ToBytesArray()?

**Comment by Zeugma440** [comment, marked as answer]
Could you please explain what you are trying to acheive ?

**Reply by Unreal-System** [reply]
I want to re-implement Icecast using csharp

**Reply by Zeugma440** [reply]
I meant, more precisely, what you are trying to acheive regarding the contents of the WAV file. 

Can you confirm you'd like to "add programmatically right header info for those memory streams without writing them to a physical file." ?

Which side of the cast are you working on ? Client-side ? server-side ? 

The SO link you provided seems to be super fine. Why can't you use it as is ?

**Reply by Unreal-System** [reply]
I am implementing server-side, but icecast does not use wav directly, only Ogg (Vorbis & Theora), Opus, WebM and MP3 streams.
The server-side needs to record the metadata of the stream when it receives the pushed stream in order to send it to the newly connected client.
Otherwise the client may not recognize the content of the stream properly.
I just tested it, Vlc Media Player/Chrome Browser can't load the new wav stream.

**Reply by Zeugma440** [reply]
> I just tested it, Vlc Media Player/Chrome Browser can't load the new wav stream

are you referring to the piece of code posted in SO ?


> The server-side needs to record the metadata of the stream when it receives the pushed stream in order to send it to the newly connected client.

Lemme rephrase that to see if I understand it correctly :
1. Server opens an audio file, reads metadata and audio data
2. Server transcodes audio data to any of the supported format (`Ogg (Vorbis & Theora), Opus, WebM and MP3`)
3. New clients connect
4. Server streams transcoded audio data (as independent streams representing chunks of the file opened in 1 ??) to all connected clients

Could you tell me at which step you'd need to use ATL ?

**Reply by Unreal-System** [reply]
Icecast doesn't work like this.
1. the server receives the stream from the administrator (here it must be in a format that the server can accept (`Ogg (Vorbis & Theora), Opus, WebM and MP3`))
2. the server caches the metadata in the stream
3. new client connects
4. the server sends the cached metadata to the new client
5. the server sends the data from the administrator's stream to all clients (the server does not cache any information other than metadata)

This is a bit of Pseudocode that explains how the server works
```
class Icecast
{
	private NetworkStream adminStream;
	private NetworkStream clientStream;
	private Metadata metaData; // typeof(byte[])
	
	public Icecast()
	{
		ThreadPool.QueueUserWorkItem(ReadAdminStreamLoop);
		ThreadPool.QueueUserWorkItem(SendMetadataStream);
		ThreadPool.QueueUserWorkItem(SendMediaStream);
	}
	
	private void ReadAdminStreamLoop()
	{
		while (true)
		{
			var bytes = adminStream.Read();
			var track = new Track(bytes);
			metaData = track.GetMetadata();
			Thread.Sleep(100);
		}
	}
	
	private void SendMetadataStream()
	{
		// new client in
		var client = new NetworkStream();
		client.Write(metaData);
	}
	
	private void SendMediaStream()
	{
		while (adminStream.Length > 0)
		{
			adminStream.CopyTo(clientStream);
		}
	}
}
```

**Reply by Zeugma440** [reply]
Alright, thanks for the detailed answer !

Which part do you need ATL for ?  Transmitting metadata in the form of a `byte[]` ?  If so, which formats are expected ? 

For instance, if we send a standalone ID3v2.4 tag, will it work ? Or do we have to send :
- Vorbis metadata for OGG & OPUS streams
- ID3v2 for MP3 streams
- [Whatever works with WebM] for WebM streams

**Reply by Unreal-System** [reply]
> * Vorbis metadata for OGG & OPUS streams
> * ID3v2 for MP3 streams
> * [Whatever works with WebM] for WebM streams

That's pretty much it, Transmitting metadata in the form of a `byte[]`.
But Icecast only provides a simple implementation (that is, it only accepts formats (`Ogg (Vorbis & Theora), Opus, WebM and MP3`)).
Other formats cannot automatically play the next stream with metadata in Vlc Media Player/Chrome Browser (i.e. the first song finishes playing, but it cannot continue with the second song, even if it receives the correct metadata).

Of course this does not mean that it cannot transfer other audio files.
It's just that transferring other audio files requires working with the client to reacquire the stream and reset the player.

**Reply by Unreal-System** [reply]
Maybe it's because I don't understand the composition of audio files, and I can only read a small amount of C code
The Metadata I understand contains at least

`Audio properties (from audio data) : Bitrate, Sample rate, Duration, VBR, Codec family, Channels count and arrangement`

After all, it is the key information that makes the audio play properly

**Reply by Zeugma440** [reply]
Looks like the IceCast protocol expects audio information in text format, not in native format

e.g. `ice-audio-info: ice-samplerate=44100;ice-bitrate=128;ice-channels=2`

See https://stackoverflow.com/questions/5215019/icecast-2-protocol-description-streaming-to-it-using-c-sharp and https://gist.github.com/ePirat/adc3b8ba00d85b7e3870

These elements can be read from any file using `Track` (see https://github.com/Zeugma440/atldotnet/wiki/3.-Usage-_-Code-snippets#base).

Is there anything else you need ?

**Reply by Unreal-System** [reply]
[Metadata is expected to be embedded into the stream. For legacy formats (MP3, AAC...) metadata can't be included in the stream, so it has to be updated out of band using the metadata update endpoint in Icecast admin to do so.](https://gist.github.com/ePirat/adc3b8ba00d85b7e3870?permalink_comment_id=2111195#gistcomment-2111195)

[@Palakis Yes, the out of band metadata are only used for "unsupported" formats like MP3 and AAC (ADTS), which have no other way to transmit metadata. For Ogg and WebM it is expected that the source client muxes the metadata into the container, and in fact, the OOB endpoint does not work for these.](https://gist.github.com/ePirat/adc3b8ba00d85b7e3870?permalink_comment_id=2988106#gistcomment-2988106)

I just did a test too, for mp3 it really does not need metadata, but Ogg (Vorbis) is needed (other formats may also need), without metadata can not be played.

**Reply by Zeugma440** [reply]
Okay, so there's extra stuff to be done, but I'd like to be sure of what is needed before starting to work on it.

There are two things that still aren't clear to me :

1/ When you say "metadata", do we agree we're still talking about `Audio properties (from audio data) : Bitrate, Sample rate, Duration, VBR, Codec family, Channels count and arrangement` ?

2/ What is the expected format ?  (don't say `byte[]`, that's just the container; I'm talking about the way these metadata should be sequenced / presented into the stream)

**Reply by Unreal-System** [reply]
1/ Yes, I agree that the "metadata" we are talking about is actually the `audio properties (from audio data) : Bitrate, Sample rate, Duration, VBR, Codec family, Channels count and arrangement`
2/ The expected format order should be the same as the original audio "metadata", but of course I don't know much about the order contained in different audio types of formats, so if it carries [Standard Metadata](https://github.com/Zeugma440/atldotnet#:~:text=count%20and%20arrangement-,Standard%20Metadata,-(from%20tags)) that's acceptable

**Reply by Zeugma440** [reply]
What we call "metadata" in 1/ isn't stored nor formatted the same way as what we call "standard metadata" in 2/. These are usually stored in two separate sections of the audio file.

However, ATL cannot format "metadata" (in the sense of 1/) as it would appear natively inside an audio file; that's why these fields are read-only in `Track`. The best ATL could do would be to give you the offset and size of that data chunk inside the audio stream, for you to be able to extract it.

Would that be okay for you ? Which audio format shall we test that on ?

**Reply by Unreal-System** [reply]
This is also possible, we can start with Ogg (Vorbis) first.
[It looks like it only needs `Identification Header, Comment Header, Setup Header`](https://xiph.org/vorbis/doc/Vorbis_I_spec.html#x1-190001.3.1)

**Reply by Zeugma440** [reply]
Alright, let's do that 👍 

I'll post some code for you to test with when there's something usable.

The goal is to come up with a generic API that allows us to handle all formats the same way.

**Reply by Zeugma440** [reply]
You're lucky, it looks like the library already has the info you need for OGG files.

- The offset of the Vorbis identification packet is always `0`, as that packet must be the first one of the file;
- The identification packet is always followed by the Vorbis comment packet, followed itself by the Vorbis setup packet ;
- The offset for the last byte of the Vorbis setup packet is given by `Track.TechnicalInformation.AudioDataOffset - 1`.

=> If you extract **bytes 0 to `Track.TechnicalInformation.AudioDataOffset - 1` (included)**, you should have what you asked for.

Please tell me if that works for you

**Reply by Unreal-System** [reply]
No need for `Track.TechnicalInformation.AudioDataOffset - 1`,extract bytes 0 to `Track.TechnicalInformation.AudioDataOffset` (included) and this will work fine!
Let's keep trying the others!

I apologize for being a little late in responding today, but the nature of my job is such that it takes up random time to work.

**Reply by Zeugma440** [reply]
Good news !! I'm glad to see it works 😄 

As for the others :
- **OPUS** : OPUS is very similar to OGG. Using the same algorithm as OGG should normally work
- **MP3** : If you need the MP3 technical header, you should extract bytes `Track.TechnicalInformation.AudioDataOffset` to `Track.TechnicalInformation.AudioDataOffset + 4`, as the header is only 4-bytes long
- WEBM : ATL currently doesn't support Matroska and its sibling formats. See #57

**Reply by Unreal-System** [reply]
Hmmm...  `Track.TechnicalInformation.AudioDataOffset` is universal for most formats?

**Reply by Zeugma440** [reply]
Yes it is, but it doesn't directly answer your needs : it marks the beginning of audio data, while you're looking for the beginning and the end of the native headers (at least that's what you got with OGG).

More information about that field here  : #90
